### DISCUSSION TITLE
How to combine `toggle-concealer` and set the `conceallevel` in one single keymap?

### DISCUSSION THREAD

**Discussion by amalgame21** [nvim-neorg/neorg#1066] [Q&A] [answered]
Sorry my lua skill is so so limited.
I have something like this in the `init.lua` file to toggle the concealer and to change `conceallevel`, so that both the actual characters of basic markups and nesting can be toggled together.

```lua
vim.api.nvim_create_autocmd({"BufEnter", "BufWinEnter"}, {
	pattern = "*.norg",
	callback = function()
		vim.opt.conceallevel = 2
		vim.keymap.set( "n", "<leader>ntc", ":Neorg toggle-concealer<CR>")
		vim.keymap.set( "n", "<leader>ntl", function()
			if vim.opt.conceallevel:get() == 2 then
				vim.opt.conceallevel = 0
			elseif vim.opt.conceallevel:get() == 0 then
				vim.opt.conceallevel = 2
			end
		end)
	end,
})
```
I would like to combine both keymaps into one, how can I do that?
I have tried ```vim.cmd(":Neorg toggle-concealer")``` right above/beneath the if-else statement but it took no effect.
Thanks!

**Comment by max397574** [comment]
I think the syntax should be without the `:` or just use `vim.cmd.Neorg("toggle-concealer")`

**Reply by amalgame21** [reply]
Thanks for your prompt response!!
Actually both `vim.cmd(":Neorg toggle-concealer")` and `vim.cmd.Neorg("toggle-concealer")` is working, but only if I comment out the if-else statement.
I do not know why, if the if-else statement just combine with one of those, the 'toggle-concealer' fail to execute.
when I set the conceallevel to other value, it also somehow toggle the concealer in a strange way.

**Reply by max397574** [reply]
it also doesn't work if you first toggle the concealer and then set the conceallevel?
if so. What if you put a ? vim.schedule(function() ... end)` around the if-statement?

**Reply by amalgame21** [reply]
I tested the vim.schedule,  same result.

However I found some strange behavior not regarding the keymaps.
In the state of concealer off, when I set `conceallevel` to any value (0 or 2 or anything else), the concealler will be VISUALLY turned on.
Then when I type `:Neorg toggle-concealer` , it visually nothing changed, still in the "on" state.
If I type the command again, concealer will be turned off.

I do not know if it is by design or not.

**Reply by max397574** [reply]
perhaps toggle-concealer does set conceallevel to 0 (or 2? whichever doesn't conceal) and then you basically reset this again
I'd have to look into the concealer code to figure this out. But I don't have time for this 2day

**Reply by amalgame21** [reply]
In my case toggle-concealer does not change conceallevel.
btw do we have a command to check the state of concealer, like we can check conceallevel with `vim.opt.conceallevel:get()` ?

Thanks for your help! Take your time!

**Reply by champignoom** [reply, marked as answer]
https://github.com/nvim-neorg/neorg/pull/1068 it's a bug

**Reply by amalgame21** [reply]
Thank! I can confirm than this keymap is working now after applying the patch:
```lua
vim.opt.conceallevel = 2
vim.keymap.set( "n", "<leader>nt", function()
	vim.schedule(function()
		if vim.opt.conceallevel:get() == 2 then
			vim.opt.conceallevel = 0
		elseif vim.opt.conceallevel:get() == 0 then
			vim.opt.conceallevel = 2
		end
	end)
	vim.cmd.Neorg("toggle-concealer")
end)

```
And the `vim.schedule()` is necessary for it to work. Thank you guys very much!

**Reply by champignoom** [reply]
@amalgame21 could you try the patch again and see if `vim.schedule()` can be omitted now?

**Reply by vhyrro** [reply]
Should be all working as intended now thanks to champignoom's fix :)

**Reply by amalgame21** [reply]
Yes `vim.shedule()` can be omitted now! thanks champignoom!
